File: 030_substring_with_concatenation_of_all_words/slidingWindow/Solution.java
import java.util.*;

class Solution {
    public List<Integer> findSubstring(String s, String[] words) {
        List<Integer> results = new ArrayList<>();
        int sLen = s.length(), arrLen = words.length;
        if (sLen == 0 || arrLen == 0) return results;
        int wordLen = words[0].length();

        Map<String, Integer> wordMap = new HashMap<>();
        for (String word : words)
            wordMap.put(word, wordMap.getOrDefault(word, 0) + 1);

        Map<String, Integer> curMap = new HashMap<>();
        int windowSize = wordLen * arrLen;
        for (int i = 0; i < wordLen; i++) {
            int start = i;
            while (start + windowSize <= sLen) {
                String str = s.substring(start, start + windowSize);
                curMap.clear();
                int j = arrLen;
                while (j > 0) {
                    String curWord = str.substring((j - 1) * wordLen, j * wordLen);
                    int count = curMap.getOrDefault(curWord, 0) + 1;
                    if (count > wordMap.getOrDefault(curWord, 0)) break;
                    curMap.put(curWord, count);
                    j--;
                }
                if (j == 0) results.add(start);
                start += wordLen * Math.max(j, 1);
            }
        }

        return results;
    }

    public static void main(String[] args) {
        String s = "barfoothefoobarman";
        String[] words = new String[]{"foo", "bar"};
        Solution solution = new Solution();
        List<Integer> results = solution.findSubstring(s, words);
        for (int result : results) {
            System.out.print(result + " ");
        }
    }
}

File: 030_substring_with_concatenation_of_all_words/slidingWindow/Solution.py
class Solution:
    def findSubstring(self, s: str, words: list) -> list:
        results = []
        s_len, arr_len = len(s), len(words)
        if s_len == 0 or arr_len == 0: return results
        word_len = len(words[0])

        word_map = {}
        for word in words:
            word_map[word] = word_map.get(word, 0) + 1

        cur_map = {}
        window_size = word_len * arr_len
        for i in range(word_len):
            start = i
            while start + window_size <= s_len:
                cur_str = s[start: start + window_size]
                cur_map.clear()
                j = arr_len
                while j > 0:
                    word = cur_str[(j - 1) * word_len: j * word_len]
                    count = cur_map.get(word, 0) + 1
                    if count > word_map.get(word, 0): break
                    cur_map[word] = count
                    j -= 1
                if j == 0: results.append(start)
                start += word_len * max(j, 1)

        return results


if __name__ == '__main__':
    s = 'barfoothefoobarman'
    words = ["foo","bar"]
    solution = Solution()
    print(solution.findSubstring(s, words))

